[libc++] Optimizations for uniform_int_distribution

LRFLEW opened the pull request:
I noticed that the implementation for `std::uniform_int_distribution` in libc++ often has worse performance than other implementations. While it might be useful to change the implementation to one that rejects less samples, such a change would result in the output of the distribution changing, which could break programs dependent on the behavior. Instead, this PR attempts to optimize the existing implementation without changing the output of the distribution.

In libc++, `std::uniform_int_distribution` uses `__independent_bits_engine` to produce n-bit random numbers. The implementation of `__independent_bits_engine` is based on the specification for `std::independent_bits_engine`, except that the value for n is specified as a runtime value. However, the design of `std::independent_bits_engine` was optimized for the case where n is known at compile time. The algorithm used requires calculating a number of values, which `std::independent_bits_engine` can calculate at compile time, avoiding any runtime performance penalty. However, the use of this algorithm for `std::uniform_int_distribution` results in these values having to be calculated for every output generated, which ends up accounting for a significant portion of the algorithm's runtime.

While analyzing the definitions of the values that need to be generated and the algorithm in general, I identified two common cases where the algorithm could be simplified:

* When R is a power of two, y0 = R, and if it's used, y1 = R. This means that the algorithm will never reject any samples generated, allowing it to use a simplified implementation of `__eval` that limits branching and doesn't read the values for y0 or y1 (allowing the optimizer to potentially eliminate the variables entirely). Additionally, the check for the condition R - y0 <= floor(y0 / n) simplifies to 0 <= floor(R / n), which is always true, meaning we can skip this check entirely when generating the required values.
* When w <= m, we know that the result of evaluating the required values will always result in n = 1. This is because, when n = ceil(w / m) = 1, R - y0 <= floor(y0 / n) can be simplified to just R <= 2\*y0, which is always true. (You can see it's always true by noting that the definitions for y0 and m gives y0 >= 2^m and R < 2^(m+1)). With this, we can provide an optimized implementation for this case that skips the multiple divisions that are required for larger values of w.

Additionally, I noticed that pretty much every use of a bit shift in the original implementation was guarded by an if clause that ensured the shift value did not cause UB. However, these checks introduce complexity and branching that can impact performance. I was able to remove these if statements by carefully considering what all the possible shift values can be. Note, I'm working with the assumption that 0 < w <= numeric_limits<result_type>::digits, which appears to be a safe assumption to make about w given how `__independent_bits_engine` is used by `std::uniform_int_distribution`.

* For computing mask0, the method used is already safe. The definition of w0 ensures that 0 < w0 <= EDt, so the value for (EDt - w0) is in the range 0 <= (EDt - w0) < EDt, which is the valid range for the shift.
* For computing y0, the current method requires checking that w0 < EDt, since it's possible that w0 = EDt. This optimization handles this case by changing the evaluation of y0 to not use any shifts, and instead use the value of mask0. This prevents the possibility of UB from an invalid shift value by simply not performing any shifts.
* For computing mask1 and y1, we have that if n > n0, then w0 + 1 <= EDt. This means that this case can be handled with the same method used for mask0 and y0. For the case where n = n0, then the values for mask1 and y1 go unused, and so we don't need to calculate them anyways. To avoid uninitialized variable warnings, this change defaults to using mask1 = mask0 and y1 = y0.
* For the shifts in `__eval`, the case where n > 1 will never result in a shift value that results in UB. This is because the relation n0\*w0 + (n - n0)\*(w0 + 1) = w gives that w0 < w, and if n > n0, w0 + 1 < w. This means the only case that needs to be handled is when n = 1. This is done by manually unrolling the k = 0 iteration and removing the shift step from it. This makes the assumptions that n > 0 and n0 > 0, but this is always true given w > 0 and n - n0 < n respectively (the latter being given by the "mod n" in the definition of n0).

All together, these changes provide minor performance improvements for sampling large ranges and significant performance improvements for sampling small ranges. The only other thing to note with this is that it removes the specific (and optimized) case for when Rp = 0 (representing R = 2^WDt) that was present previously. However, the optimizations for power-of-two values for R apply to this case as well, and there doesn't appear to be much benefit to trying to further optimize this case. This change also reuses the overload for `__eval` previously used for the Rp = 0 special case for handling power-of-two values of R.

LRFLEW committed:
Update Tests for uniform_int_distribution

LRFLEW commented:
I included two test files I made while validating this change as part of this PR.

The first is a test file that tests the output of `std::uniform_int_distribution` against lists of expected outputs. The lists were generated using LLVM 19.1.7, and are intended as a means of validating that the changes I made didn't change the output of the distribution. Since the C++ standard doesn't specify how distributions are supposed to generate values, this test does not validate standard conformance at all, and only tests if the output has changed relative to previous LLVM versions.

The second is a benchmark for `std::uniform_int_distribution`. It tests the distribution using minstd_rand0 and ranlux24_base (chosen for their speed, which minimizes the impact of the PRNG performance on the measured time), which tests both when R isn't and is a power of two respectively. Both engines are tested using three different sample ranges: 2^20, chosen to result in n = 1, 2^40, chosen to result in n = 2 and n0 = 2, and 2^41, chosen to result in n = 2 and n0 = 1. This gives good coverage of the different possible results based on my optimization efforts.

I ran this test on my machine (i7 8700K, WSL on Windows 10) with and without the changes made in this PR, took the median time of 16 repetitions, and compiled the result in the table below. This should give a good idea of the impact this change will have.

| Engine and Sample Size | Original Time | New Time | Performance |
| --- | --- | --- | --- |
| minstd_rand0, w = 20 | 39.3 ns | 13.9 ns | 2.83x |
| ranlux24_base, w = 20 | 47.9 ns | 23.0 ns | 2.08x |
| minstd_rand0, w = 40 | 45.3 ns | 38.8 ns | 1.17x |
| ranlux24_base, w = 40 | 61.3 nss | 46.7 ns | 1.31x |
| minstd_rand0, w = 41 | 45.8 ns | 40.3 ns | 1.14x |
| ranlux24_base, w = 41 | 61.2 ns | 48.7 ns | 1.26x |

I kept these test file additions in its own commit preceding the changes to the libc++ header. I did this because a) it makes it easier to perform the before-and-after tests, and b) I'm not sure if these test files should be merged. While the benchmark *may* be useful, the other test file likely isn't useful, since it doesn't test anything related to standard conformance or a known regression. I'll leave the tests as they are now for code review. However, let me know if you want me to squash the commits or remove any of the test files, and I'll do it.

LRFLEW committed:
Optimize uniform_int_distribution

LRFLEW commented:
> Do you think we should use a different algorithm that has improved performance?

Short Answer: Maybe, but what to replace it *with* is a non-trivial problem. Also, this PR addresses a lot of the concern I have with the current method.

For context, I'm going to reference the algorithms as described in [this blog post on the PCG website](https://www.pcg-random.org/posts/bounded-rands.html), because it's one of the best write-ups of the various resampling methods I've found. Note, however, that the code snippets (and to a lesser extent, the algorithms themselves) use a much more narrow contract than `std::uniform_int_distribution`. The algorithms on that website are all narrowing (i.e. the output range is smaller (or equal to) the PRNG's range) and the PRNG is assumed to be a 32-bit PRNG (i.e. the output range is the full range of unsigned 32-bit integers). In contrast, `std::uniform_int_distribution` needs to be able to support any possible input or output range.

The algorithm used here is basically the "Bitmask" method described there. This method has a critical trade-off, where it avoids having to do a more expensive division (or multiplication), but it generally performs significantly more sample rejections. Whether this performs better or worse depends on multiple factors, such as how fast the PRNG algorithm is, how fast division is on the device, whether the CPU has a CLZ instruction (or if it has to be emulated in software), and what kinds of ranges are actually sampled.

There is one more thing to keep in mind with this though: the "Bitmask" method assumes the PRNG is a power-of-two PRNG. Since `std::uniform_int_distribution` can't make that assumption, it has to coerce the result of the PRNG into a binary range. This means that non-Po2 PRNGs, including minstd, have to go through two rejection samples, with further impacts performance. Plus, the method used doesn't seem to be particularly optimized for runtime performance. This PR primarily addresses the runtime performance of this step, which helps a lot with closing the performance gap where it occurs.

There are a lot of alternative options if we did want to change the algorithm. From what I've seen, libstdc++ uses a combination of the "Debiased Integer Multiplication — Lemire's Method" and "Division with Rejection (Unbiased)" methods depending on the PRNG and available instruction set. There's also the different modulus methods, which might be worth considering. It's also worth keeping in mind that the current method provides the widening method (where the output range is larger than the input range) basically for free, while switching methods will require deciding on a method for that as well.

If there's interest in possibly changing the algorithm used, I'd be open to having and participating in a larger discussion around it. It should probably include performing performance tests in a wider range of hardware than I personally have access to, as we probably shouldn't just make assumptions about all hardware based on modern high-end hardware. In the meantime, I think it's worth considering this PR on its own, as it addresses a lot of the concern I have with the current method.

ldionne commented on libcxx/test/benchmarks/numeric/rand.uni.int.bench.cpp:
```diff
@@ -0,0 +1,36 @@
+//
+// UNSUPPORTED: c++03
+
+#include <benchmark/benchmark.h>
+#include <cstdint>
+#include <random>
```
```suggestion
#include <cstdint>
#include <random>

#include <benchmark/benchmark.h>
```

Nitpick!

ldionne commented on libcxx/test/benchmarks/numeric/rand.uni.int.bench.cpp:
```diff
@@ -0,0 +1,36 @@
+#include <random>
+
+template <typename Eng, std::uint64_t Max>
+static void bm_rand_uni_int(benchmark::State& state) {
```
Let's name this `bm_uniform_int_distribution` so the benchmark names are clearer when running the benchmark. WDYT?

LRFLEW replied:
That works for me. I had just used this name to match the file's name, but I agree that using the full class name makes more sense here.

ldionne commented on libcxx/include/__random/uniform_int_distribution.h:
```diff
@@ -64,7 +64,7 @@ class __independent_bits_engine {
   _LIBCPP_HIDE_FROM_ABI __independent_bits_engine(_Engine& __e, size_t __w);
```
In your PR description, you mention:

> Note, I'm working with the assumption that `0 < w <= numeric_limits<result_type>::digits`, which appears to be a safe assumption to make about w given how `__independent_bits_engine` is used by `std::uniform_int_distribution`.

We should add an internal assertion like `_LIBCPP_ASSERT_INTERNAL(0 < w <= numeric_limits<result_type>::digits, "explain why this is needed")`, that way we'll catch it if our implementation has a bug. I am assuming that this assumption can't fail based on user input, right?

ldionne commented on libcxx/test/std/numerics/rand/rand.dist/rand.dist.uni/rand.dist.uni.int/output.pass.cpp:
```diff
@@ -0,0 +1,653 @@
+//===----------------------------------------------------------------------===//
```
I think it's great that you added this test and documented it in your PR. That's the correct methodology for this kind of change. However, as you correctly mentioned in a comment, I think we shouldn't check-in the test since its value outside of this change isn't compelling.

ldionne commented on libcxx/include/__random/uniform_int_distribution.h:
```diff
@@ -64,7 +64,7 @@ class __independent_bits_engine {
   _LIBCPP_HIDE_FROM_ABI __independent_bits_engine(_Engine& __e, size_t __w);
 
```
Unrelated to this file: let's add a release note for this optimization. You can add it in `libcxx/docs/ReleaseNotes/21.rst`.

LRFLEW replied:
Do you have a suggestion on how to word this exactly? Should I just put something like this under "Improvements and New Features"?

> The implementation of `std::uniform_int_distribution` has been optimized for common cases.

ldionne reviewed (commented):
First of all, thanks for the amazingly detailed writeup. This is very useful to vet this change since most of us are not super familiar with the details of our `<random>` implementation, which was written early in the library's lifetime and not touched much afterwards.

I will echo @philnik777 's comment that changing the exact values returned by this type is not a problem, as long as we retain a similarly good (or better) uniform distribution. Changing the ABI of the type is impossible to do *by default*, however we can definitely do it behind [an ABI macro](https://libcxx.llvm.org/ABIGuarantees.html) if there is a sufficiently large benefit. That way, adopters of unstable ABI features can opt into the change, and ABI-stable vendors creating new ABIs (typically when introducing a new architecture) can also benefit from it.

About this change itself, I pulled your patch and ran the benchmark on my machine (arm64 mac studio). The results are:

```
$ libcxx/utils/libcxx-compare-benchmarks build/default build/candidate libcxx/test/benchmarks/numeric/rand.uni.int.bench.cpp
Comparing build/default/libcxx/test/benchmarks/numeric/Output/rand.uni.int.bench.cpp.dir/benchmark-result.json to build/candidate/libcxx/test/benchmarks/numeric/Output/rand.uni.int.bench.cpp.dir/benchmark-result.json
Benchmark                                                         Time             CPU      Time Old      Time New       CPU Old       CPU New
----------------------------------------------------------------------------------------------------------------------------------------------
bm_rand_uni_int<std::minstd_rand0, 1ull << 20>                 -0.0434         -0.0434             7             7             7             7
bm_rand_uni_int<std::ranlux24_base, 1ull << 20>                -0.1183         -0.1183            10             9            10             9
bm_rand_uni_int<std::minstd_rand0, 1ull << 40>                 -0.0177         -0.0176            13            12            13            12
bm_rand_uni_int<std::ranlux24_base, 1ull << 40>                -0.1081         -0.1081            16            14            16            14
bm_rand_uni_int<std::minstd_rand0, 1ull << 41>                 -0.1236         -0.1246            14            13            14            13
bm_rand_uni_int<std::ranlux24_base, 1ull << 41>                -0.1553         -0.1561            17            15            17            15
OVERALL_GEOMEAN                                                -0.0957         -0.0959             0             0             0             0
```

So overall, I'm seeing roughly 10% improvement, which is great. Do you have an hypothesis to explain why you were seeing up to 2.83x on some benchmarks and I am seeing more modest numbers?

I think the next step for landing this patch would be to address my minor review comments, rebase this (there are no conflicts) and this is probably fine to land. Any further improvement in the algorithm or potential ABI breaking change should be discussed in a follow-up. Let's not make perfection the enemy of good!

LRFLEW commented:
Ok, I've went and addressed (most) of the review comments and rebased against the main branch. I removed the output test case, since that was only intended to be temporary anyways, but kept the benchmark on its own commit for a bit. The testing results you got has me wanting to try running the benchmark on some other devices to try and get a more complete picture of what the results of this change are.

> So overall, I'm seeing roughly 10% improvement, which is great. Do you have an hypothesis to explain why you were seeing up to 2.83x on some benchmarks and I am seeing more modest numbers?

The large improvements I saw were for the case where n = 1, with the main optimization there being the removal of two division calculations (to calculate the values for `n` and `w0`). I suspected maybe Apple put some sort of optimization in their silicon for trivial divisions (eg. `x / 1 = x` and `x / x = 1`), but looking into it, apparently [Apple Silicon can just perform integer division way faster than Intel](https://ridiculousfish.com/blog/posts/benchmarking-libdivide-m1-avx512.html). This result is a surprise to me, and means that Apple computers aren't going to see as much of an impact from this. I kind of want to see the benchmark results from another ARM64 CPU (to see if it's just Apple's CPUs), such as a Raspberry Pi, but I don't have one of those on hand to test it with.
